[14.0][TMP] fix to resolve calendar conflict with DDMRP

mmequignon opened the pull request:
There's a conflict with DDMRP regarding the usage of the
stock.warehouse's calendar_id.
This PR implements a new `calendar2_id` field (already there in v13, see
this commit https://github.com/OCA/sale-workflow/pull/1696/commits/5b0b93a126706b3f32e789b6e751099001b16bcc)
in order to temporary solve this conflict.

mmequignon committed:
[TMP] fix to resolve calendar conflict with DDMRP

There's a conflict with DDMRP regarding the usage of the
stock.warehouse's calendar_id.
This PR implements a new `calendar2_id` field (already there in v13, see
this commit https://github.com/OCA/sale-workflow/pull/1696/commits/5b0b93a126706b3f32e789b6e751099001b16bcc)
in order to temporary solve this conflict.

sebalix commented:
The commit in the description belongs to the following PR in 13.0: https://github.com/OCA/sale-workflow/pull/1696

sebalix committed:
[FIX] sale_order_line_date: fix commitment_date onchange

In 14.0 there are two methods named `_onchange_commitment_date':

1) one which set the `commitment_date` with the `expected_date` value
   (triggered by `expected_date`)
2) the second which triggers a warning (triggered by `commitment_date`
   only)

The first one is never executed because the second has the precedence,
and it is a good thing because we do not want to automatically set the
`commitment_date` by default!
Odoo has fixed this code in 15.0 by removing the first method and added
the `expected_date` field as trigger on the second one
(see 4031e86f2564378d18f7a930ad45f178c71b2939).

This commit is adding `expected_date` as well on this overridden method
so we are in line with the expected behavior.

sebalix committed:
[FIX] sale_delivery_date: add dependency

Add `sale_order_line_date` as dependency because this addon is
overriding the same methods (`_onchange_commitment_date` and
`_prepare_procurement_values`) and we want to ensure than
`sale_delivery_date` takes the precedence over `sale_order_line_date`.

Also, the other feature of `sale_order_line_date` which sync the
`commitment_date` between SO and its lines is welcome in
`sale_delivery_date`.

sebalix commented:
Added the same fixup taken from https://github.com/OCA/sale-workflow/pull/1696 , but it'll still need to be improved to be TZ compliant as mentionned by @jbaudoux here: https://github.com/OCA/sale-workflow/pull/1696/files#r739055759=

sebalix committed:
[FIX] sale_delivery_date: take into account 'date_deadline' field

sebalix commented:
PR based on https://github.com/OCA/sale-workflow/pull/1968 to resolve conflicts while including fixes.
Anyway it's a temporary branch.

sebalix committed:
[FIX] sale_delivery_date: find a working day matching the customer's delivery time window

sebalix committed:
[FIX] sale_delivery_date: fix computation of the expected date

mmequignon committed:
[IMP] sale_delivery_date: convert methods to @api.model

mmequignon committed:
[IMP] sale_delivery_date: Recompute backorders dates

mmequignon committed:
[IMP] sale_delivery_date: Postpone dates if sale line qty is modified

mmequignon committed:
sale_delivery_date: reset hour before applying time window
